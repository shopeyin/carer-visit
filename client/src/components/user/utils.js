import axios from "axios";
import { BASE_URL } from "../../App";

export const addVisitInfo = async (data) => {
  try {
    await axios.post(`${BASE_URL}/visitInformation/`, data);
    console.log("submitted");
  } catch (error) {
    console.log(error);
  }
};

export const fetchVisit = async (currentUserId, visitDate) => {
  console.log(axios);
  try {
    const visitData = await axios.post(
      `${BASE_URL}/visit/${currentUserId}`,
      visitDate
    );
    const {
      data: {
        data: { visit },
      },
    } = visitData;

    console.log(visit);
    return visit;
  } catch (err) {
    return err;
  }
};
