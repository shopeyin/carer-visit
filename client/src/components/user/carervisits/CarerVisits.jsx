import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import DatePicker from "react-datepicker";
import { format } from "date-fns";
import { BASE_URL } from "../../../App";
import Visit from "./Visit";
import axios from "axios";
import { fetchVisit } from "../utils";

function CarerVisits({ currentUser }) {
  const [serviceUsersVisit, setServiceUsersVisit] = useState([]);
  const [selectedDate, setSelectedDate] = useState(new Date());

  useEffect(() => {
    let mounted = true;

    let visitDate = {
      dateOfVisit:
        format(new Date(selectedDate), "yyyy-MM-dd") + "T00:00:00.000+00:00",
    };

    const fetchVisitData = async () => {
      let visit = await fetchVisit(currentUser._id, visitDate);
      console.log(visit); //prototype = promise, promise state = fulfilled, promise Result = [1,2,3]
      // visit.then(v=>console.log(v))

      if (visit.length === 0) {
        setServiceUsersVisit([]);
        return;
      }
      if (mounted) {
        localStorage.setItem("visitId", visit[0]._id);

        setServiceUsersVisit(visit[0].serviceusersToVisit);
      }
    };
    fetchVisitData();
    // const fetchVisit = async () => {
    //   try {
    //     const visitData = await axios.post(
    //       `${BASE_URL}/visit/${currentUser._id}`,
    //       visitDate
    //     );
    //     console.log(visitData);
    //     const {
    //       data: {
    //         data: { visit },
    //       },
    //     } = visitData;

    //     console.log(visit);

    //     if (mounted) {
    //       localStorage.setItem("visitId", visit[0]._id);

    //       setServiceUsersVisit(visit[0].serviceusersToVisit);
    //     }
    //   } catch (err) {
    //     setServiceUsersVisit([]);
    //     console.log(err.message);
    //   }
    // };

    // fetchVisit();

    return () => {
      mounted = false;
    };
  }, [currentUser, selectedDate]);

  let itemsToRender;

  if (serviceUsersVisit) {
    itemsToRender = serviceUsersVisit.map((serviceUser) => {
      return (
        <div
          className="row mt-3 pl-4 pr-4 d-flex  justify-content-center"
          key={serviceUser._id}
        >
          <div className="col-8 col-sm-5 col-md-4 col-lg-3">
            <Link to={`activities/${serviceUser._id}`} className="link-color">
              <div className="card ">
                <div className="card-body ">
                  <h4 className="card-title link-color"> {serviceUser.name}</h4>

                  <h6 className="card-text link-color ">
                    {serviceUser.address}
                  </h6>
                </div>
              </div>
            </Link>{" "}
          </div>
        </div>
      );
    });
  }
  return (
    <div className="container-fluid p-0 ">
      <div className="row  d-flex  justify-content-center mt-4 ">
        <div className="col-8 col-sm-4  text-center">
          {" "}
          <DatePicker
            selected={selectedDate}
            onChange={(date) => setSelectedDate(date)}
            dateFormat="yyyy/MM/dd"
            minDate={new Date()}
            showYearDropdown
            scrollableMonthYearDropdown
          />
        </div>
      </div>
      <Visit
        serviceUsersVisit={serviceUsersVisit}
        itemsToRender={itemsToRender}
        selectedDate={selectedDate}
      />
    </div>
  );
}

export default CarerVisits;
