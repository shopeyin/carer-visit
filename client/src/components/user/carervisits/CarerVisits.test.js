import { shallow } from "enzyme";
import React from "react";
import CarerVisits from "./CarerVisits";
import shallowToJson from "enzyme-to-json";
import { fetchVisit } from "../utils";

import mockAxios from "axios";

jest.mock("axios");

describe("carer visit component", () => {
  let mockCurrentUser = {
    role: "carer",
    _id: 1,
    name: "belinda",
    barePassword: "123456",
    email: "belinda@gmail.com",
    password: "12345666665667",
  };

  const wrapper = shallow(<CarerVisits currentUser={mockCurrentUser} />);

  it("expect to render carervisit component", () => {
    //expect(shallowToJson(wrapper)).toMatchSnapshot();

    expect(wrapper.find(".container-fluid").length).toBe(1);
  });

  // it("calls fetch visit api with accurate data", async () => {
  //   expect.assertions(1);

  //   let visitDate = {
  //     dateOfVisit: "2022-08-17T00:00:00.000+00:00",
  //   };

  //   let result = await fetchVisit("62bc001b2105fc4bd8981d70", visitDate);
  //   console.log(result);

  //   expect(result.length).toBe(1);
  // });

  // it("calls fetch visit api with inaccurate data", async () => {
  //   let visitDate = {
  //     dateOfVisit: "2022-08-17T00:00:00.000+00:00",
  //   };

  //   let result = await fetchVisit("62bc001b2105fc4bd8981d78", visitDate);

  //   expect(result.length).toBe(0);
  // });

  it("should return visit for the day", async () => {
    const visit = [
      {
        careruser: "62bc001b2105fc4bd8981d70",
        dateOfVisit: "2022-08-16T00:00:00.000Z",
        _id: "62fb259fed320240f851ebb9",
        serviceusersToVisit: [[Object]],
      },
    ];

    let visitDate = {
      dateOfVisit: "2022-08-17T00:00:00.000+00:00",
    };

    mockAxios.post.mockImplementation(() => Promise.resolve({ data: visit }));
    let result = await fetchVisit("62bc001b2105fc4bd8981d70", visitDate);
    console.log(result);
  });
});

// let wrapper;

// beforeEach(() => {
//   const mockProps = {
//     currentUser: "ola",
//   };
//   wrapper = shallow(<CarerVisits {...mockProps} />);
// });

// it("expect to render CarerVisit component", () => {
//   expect(shallowToJson(wrapper)).toMatchSnapshot();
// });
