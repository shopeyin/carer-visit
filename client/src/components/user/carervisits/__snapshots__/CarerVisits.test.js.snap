// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`carer visit component calls fetch visit api 1`] = `
<div
  className="container-fluid p-0 "
>
  <div
    className="row  d-flex  justify-content-center mt-4 "
  >
    <div
      className="col-8 col-sm-4  text-center"
    >
       
      <r
        allowSameDay={false}
        customTimeInput={null}
        dateFormat="yyyy/MM/dd"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        minDate={2022-08-16T07:49:43.396Z}
        monthsShown={1}
        nextMonthAriaLabel="Next Month"
        nextMonthButtonLabel="Next Month"
        nextYearAriaLabel="Next Year"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthAriaLabel="Previous Month"
        previousMonthButtonLabel="Previous Month"
        previousYearAriaLabel="Previous Year"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        scrollableMonthYearDropdown={true}
        selected={2022-08-16T07:49:43.396Z}
        selectsDisabledDaysInRange={false}
        shouldCloseOnSelect={true}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearDropdown={true}
        showYearPicker={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
    </div>
  </div>
  <Visit
    itemsToRender={Array []}
    selectedDate={2022-08-16T07:49:43.396Z}
    serviceUsersVisit={Array []}
  />
</div>
`;

exports[`carer visit component expect to render carervisit component 1`] = `
<div
  className="container-fluid p-0 "
>
  <div
    className="row  d-flex  justify-content-center mt-4 "
  >
    <div
      className="col-8 col-sm-4  text-center"
    >
       
      <r
        allowSameDay={false}
        customTimeInput={null}
        dateFormat="yyyy/MM/dd"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        minDate={2022-08-15T20:21:31.833Z}
        monthsShown={1}
        nextMonthAriaLabel="Next Month"
        nextMonthButtonLabel="Next Month"
        nextYearAriaLabel="Next Year"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthAriaLabel="Previous Month"
        previousMonthButtonLabel="Previous Month"
        previousYearAriaLabel="Previous Year"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        scrollableMonthYearDropdown={true}
        selected={2022-08-15T20:21:31.496Z}
        selectsDisabledDaysInRange={false}
        shouldCloseOnSelect={true}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearDropdown={true}
        showYearPicker={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
    </div>
  </div>
  <Visit
    itemsToRender={Array []}
    selectedDate={2022-08-15T20:21:31.496Z}
    serviceUsersVisit={Array []}
  />
</div>
`;

exports[`carer visit component get carer visit 1`] = `
<div
  className="container-fluid p-0 "
>
  <div
    className="row  d-flex  justify-content-center mt-4 "
  >
    <div
      className="col-8 col-sm-4  text-center"
    >
       
      <r
        allowSameDay={false}
        customTimeInput={null}
        dateFormat="yyyy/MM/dd"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        minDate={2022-08-16T06:03:53.144Z}
        monthsShown={1}
        nextMonthAriaLabel="Next Month"
        nextMonthButtonLabel="Next Month"
        nextYearAriaLabel="Next Year"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthAriaLabel="Previous Month"
        previousMonthButtonLabel="Previous Month"
        previousYearAriaLabel="Previous Year"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        scrollableMonthYearDropdown={true}
        selected={2022-08-16T06:03:53.144Z}
        selectsDisabledDaysInRange={false}
        shouldCloseOnSelect={true}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearDropdown={true}
        showYearPicker={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
    </div>
  </div>
  <Visit
    itemsToRender={Array []}
    selectedDate={2022-08-16T06:03:53.144Z}
    serviceUsersVisit={Array []}
  />
</div>
`;
